File: app/src/main/java/org/xdty/callerinfo/exporter/Exporter.java
package org.xdty.callerinfo.exporter;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.google.gson.Gson;

import org.xdty.callerinfo.BuildConfig;
import org.xdty.callerinfo.model.database.Database;
import org.xdty.callerinfo.model.database.DatabaseImpl;
import org.xdty.callerinfo.model.db.Caller;
import org.xdty.callerinfo.model.db.InCall;
import org.xdty.callerinfo.model.db.MarkedRecord;

import java.util.List;
import java.util.Map;

public final class Exporter {

    private static Gson gson = new Gson();
    private SharedPreferences mPrefs;
    private SharedPreferences mWindowPrefs;
    private Database mDatabase;

    public Exporter(Context context) {
        context = context.getApplicationContext();
        mPrefs = PreferenceManager.getDefaultSharedPreferences(context);
        mWindowPrefs = context.getSharedPreferences("window", Context.MODE_PRIVATE);
        mDatabase = DatabaseImpl.getInstance();
    }

    public String export() {
        Data data = new Data();
        data.setting = mPrefs.getAll();
        data.window = mWindowPrefs.getAll();
        data.timestamp = System.currentTimeMillis();
        data.versionCode = BuildConfig.VERSION_CODE;
        data.versionName = BuildConfig.VERSION_NAME;
        data.callers = mDatabase.fetchCallersSync();
        data.inCalls = mDatabase.fetchInCallsSync();
        data.markedRecords = mDatabase.fetchMarkedRecordsSync();
        return gson.toJson(data);
    }

    public void fromString(String s) {
        Data data = gson.fromJson(s, Data.class);
        for (Map.Entry<String, ?> entry : data.setting.entrySet()) {
            Object value = entry.getValue();
            addPref(mPrefs, entry.getKey(), value);
        }
        for (Map.Entry<String, ?> entry : data.window.entrySet()) {
            Object value = entry.getValue();
            addPref(mWindowPrefs, entry.getKey(), value);
        }
        mDatabase.addCallers(data.callers);
        mDatabase.addInCallers(data.inCalls);
        mDatabase.addMarkedRecords(data.markedRecords);
    }

    private void addPref(SharedPreferences prefs, String key, Object value) {
        if (value instanceof String) {
            prefs.edit().putString(key, (String) value).apply();
        } else if (value instanceof Integer) {
            prefs.edit().putInt(key, (Integer) value).apply();
        } else if (value instanceof Long) {
            prefs.edit().putLong(key, (Long) value).apply();
        } else if (value instanceof Float) {
            prefs.edit().putFloat(key, (Float) value).apply();
        } else if (value instanceof Boolean) {
            prefs.edit().putBoolean(key, (Boolean) value).apply();
        }
    }

    public static class Data {
        Map<String, ?> setting;
        Map<String, ?> window;
        List<Caller> callers;
        List<InCall> inCalls;
        List<MarkedRecord> markedRecords;
        String versionName;
        int versionCode;
        long timestamp;
    }

}

File: app/src/main/java/org/xdty/callerinfo/exporter/Exporter.kt
package org.xdty.callerinfo.exporter

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import com.google.gson.GsonBuilder
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import org.xdty.callerinfo.BuildConfig
import org.xdty.callerinfo.R
import org.xdty.callerinfo.model.database.Database
import org.xdty.callerinfo.model.database.DatabaseImpl.Companion.instance
import org.xdty.callerinfo.model.db.Caller
import org.xdty.callerinfo.model.db.InCall
import org.xdty.callerinfo.model.db.MarkedRecord

class Exporter(context: Context) {
    private val mPrefs: SharedPreferences
    private val mWindowPrefs: SharedPreferences
    private val mDatabase: Database
    private val uidKey: String
    fun export(): Observable<String> {
        return Observable.create<String> { emitter ->
            emitter.onNext(exportSync())
            emitter.onComplete()
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun exportSync(): String {
        val data = Data()
        data.setting = mPrefs.all
        data.window = mWindowPrefs.all
        data.timestamp = System.currentTimeMillis()
        data.versionCode = BuildConfig.VERSION_CODE
        data.versionName = BuildConfig.VERSION_NAME
        data.callers = mDatabase.fetchCallersSync()
        data.inCalls = mDatabase.fetchInCallsSync()
        data.markedRecords = mDatabase.fetchMarkedRecordsSync()
        return gson.toJson(data)
    }

    fun fromString(s: String?): Observable<String> {
        return Observable.create<String> { emitter ->
            emitter.onNext(fromStringSync(s)!!)
            emitter.onComplete()
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun fromStringSync(s: String?): String? {
        var res: String? = null
        try {
            val data = gson.fromJson(s, Data::class.java)
            for ((key, value1) in data.setting!!) {
                val value = value1!!
                addPref(mPrefs, key, value)
            }
            for ((key, value1) in data.window!!) {
                val value = value1!!
                addPref(mWindowPrefs, key, value)
            }
            mDatabase.addCallers(data.callers!!)
            mDatabase.addInCallers(data.inCalls!!)
            mDatabase.addMarkedRecords(data.markedRecords!!)
        } catch (e: Exception) {
            e.printStackTrace()
            res = e.message
        }
        return res
    }

    private fun addPref(prefs: SharedPreferences, key: String, value: Any) { // ignore uid
        if (key == uidKey) {
            return
        }
        if (value is String) {
            prefs.edit().putString(key, value).apply()
        } else if (value is Int) {
            prefs.edit().putInt(key, value).apply()
        } else if (value is Long) {
            prefs.edit().putLong(key, value).apply()
        } else if (value is Float) {
            prefs.edit().putFloat(key, value).apply()
        } else if (value is Boolean) {
            prefs.edit().putBoolean(key, value).apply()
        }
    }

    class Data {
        internal var setting: Map<String, *>? = null
        internal var window: Map<String, *>? = null
        internal var callers: List<Caller>? = null
        internal var inCalls: List<InCall>? = null
        internal var markedRecords: List<MarkedRecord>? = null
        internal var versionName: String? = null
        internal var versionCode = 0
        internal var timestamp: Long = 0
    }

    companion object {
        private val gson = GsonBuilder().setPrettyPrinting().create()
    }

    init {
        val context = context.applicationContext
        mPrefs = PreferenceManager.getDefaultSharedPreferences(context)
        mWindowPrefs = context.getSharedPreferences("window", Context.MODE_PRIVATE)
        mDatabase = instance
        uidKey = context.getString(R.string.uid_key)
    }
}